### ISSUE TITLE
Non Packed Dataset Format?

### ISSUE THREAD

**Issue by mallorbc** [huggingface/trl#637]
If you leave packing to false, we individually tokenize each item in a dataset without padding and truncating results longer than the max a tokenizer allows

https://github.com/huggingface/trl/blob/3b2c820db68aecca36c2ff5d152de59deec8fe5c/trl/trainer/sft_trainer.py#L287-L295

I printed out the input_ids of dataset after tokenization and as expected,  the length of the vectors varies based on what the text was.

However, batching still works which I do not understand. How are we stacking vectors of different sizes?

In the past, to get this to work, I added padding to the left side of the input.  

Is this still happening, just further hidden in the code or am I missing something?

Basically, I want to be sure that I am not either folding smaller inputs into the full context window(which I am pretty sure is not the case because that is what packing does), and I want to be sure that the text input ids are not being cut off.

Thanks!

**Comment by lvwerra** [comment]
So the `SFTTrainer` uses a default data collator which is applied by default to the processed dataset. It pads the sequences in a batch to the max sequence length of that batch on the fly. You can overwrite it by passing your own `data_collator`. Hope that clarifies things!

**Comment by younesbelkada** [comment]
Closing for now, feel free to re-open if you need more help !
